### PR TITLE
Early deprecate functions in DoFTools

### PR DESCRIPTION
**Pull request by peterrum**
... many functions already return the `IndexSet`s directly.

### CODE REVIEW

**Comment by kronbichler** [comment]
I am not particularly happy about the incompatibility caused by this change. Even though this is only an early deprecation, we will soon get tons of warnings and then error for almost every single deal.II program that uses MPI. I see the point in terms of some other functions in `DoFTools`, so the interface makes sense and I support it. But I reserve myself against the early deprecation warning already at this stage: I think it should come only when we have gone through all functions with similar interfaces (uniformly return things computed inside function). I think one of the worst things we can do to our users is to randomly deprecate some of the functions for every release, rather than providing a (somewhat) uniform solution which covers most places.

**Comment by peterrum** [comment]
> I am not particularly happy about the incompatibility caused by this change. Even though this is only an early deprecation, we will soon get tons of warnings and then error for almost every single deal.II program that uses MPI. I see the point in terms of some other functions in DoFTools, so the interface makes sense and I support it. But I reserve myself against the early deprecation warning already at this stage: I think it should come only when we have gone through all functions with similar interfaces (uniformly return things computed inside function). I think one of the worst things we can do to our users is to randomly deprecate some of the functions for every release, rather than providing a (somewhat) uniform solution which covers most places.

I have dropped the pragma but would suggest to keep the comment regarding deprecation. Is that fine?

File: examples/step-32/step-32.cc [hunk 1]
```diff
@@ -1864,16 +1864,16 @@ namespace Step32
       stokes_partitioning.push_back(stokes_index_set.get_view(0, n_u));
       stokes_partitioning.push_back(stokes_index_set.get_view(n_u, n_u + n_p));
 
-      DoFTools::extract_locally_relevant_dofs(stokes_dof_handler,
-                                              stokes_relevant_set);
+      stokes_relevant_set =
+        DoFTools::extract_locally_relevant_dofs(stokes_dof_handler);
```
Comment (kronbichler) [inline examples/step-32/step-32.cc:1868]: I think in this place the regular RVO will probably not help the change will give rise to (very slightly) slower code than we had before, as we need to call the `operator=` here from the temporary object. This is different from the case in `step-18` you made above where RVO eliminates the copy (on sane compilers).

I think this is justified for a cheap object like `IndexSet` given the interface uniformity it brings us. Furthermore, I believe we should at some point make the `locally_relevant_dofs` a member variable of the `DoFHandler`, given how often we set up and use these objects and how little memory they consume in the grand scheme of things. Then this gets even more clear that we should do it. But I do not think the case is necessarily as low for all functions that *merely fill* something, so we might need to make a case-by-case decision.

**Review by kronbichler — APPROVED** [review]

**Review by tamiko — APPROVED** [review]

File: examples/step-32/step-32.cc (hunk 1 above)
Comment (bangerth) [reply examples/step-32/step-32.cc:1868]: See also #11896.

**Comment by bangerth** [comment]
Yes, nice. I like these functions a lot better and had wanted to make this change myself for a long time. Thanks for beating me to it!

**Pull request merged**